### ISSUE TITLE
Line length limit in fourmolu

### ISSUE THREAD

**Issue by gustavoavena** [fourmolu/fourmolu#71]
**TL;DR:** Would an option for max number of characters in a line be a welcome feature in fourmolu? Is anyone asking about or working on this?

Based on ormolu's design principle's and a couple of comments that I've seen in its repo (e.g. [this](https://github.com/tweag/ormolu/issues/508)), I know that setting a limit to the number of character in a line is not something that they would want to do. 

FWIU, fourmolu was forked to provide more customization, so it now provides multiple options which I imagine wouldn't be considered in ormolu. With this in mind, I wanted to ask the maintainers about possibly adding an option for max number of characters in a line. 

- Would this be a welcome feature? Or is there a reason not to do this (e.g. increases the separation from ormolu, harder to maintain)?
- Has anyone asked for this before? Is anyone working or planning to work on this already?

Thanks!

**Comment by georgefst** [comment]
> Would this be a welcome feature?

Pretty much any feature is welcome here! Unless it would be a real pain to maintain for any reason.

> Has anyone asked for this before?

Kind of: #19 

> Is anyone working or planning to work on this already?

No. I would like to see it, but it's fairly low on my list in terms of priority.

**Comment by georgefst** [comment]
I had a hunch that this might be a more invasive change than anything else we've added, but it's hard to know without just trying to implement it.

**Comment by parsonsmatt** [comment]
Sure! 

I'll even go against my own preferences here and allow people to customize it. As much as I might prefer the 80 column limit, I know other people prefer more (and sometimes less) ;)

**Comment by gustavoavena** [comment]
Thank you both for the quick answers! I'm happy to hear that features like these would be welcome! 🙂

Yeah, I think that allowing users to set a max length would be a significant change... I'm not sure if I'll be able to work on something this big, but there might be some small tweaks that could achieve similar results (i.e. getting rid of very long lines and making the code easier to read).

FWIU, fourmolu is already able to break lines properly in things like lists or explicit import lists. The "issue" is that the user needs to break one of these lines for fourmolu to break and align the rest.

Maybe providing an option to enable automatic line-breaking for these cases (e.g. explicit imports, lists) could be a good start that would cover most of the long lines... WDYT?

My initial idea was to break lists with multiple elements if this option is enabled. Something like:

### Before
```haskell
import Foo (foo)
import Bar (bar1, bar2, bar3)

singleElementList :: [String]
singleElementList = ["one string"]

multipleElementsList :: [String]
multipleElementsList = ["one", "two", "three"]

```

### After
```haskell
import Bar (
  bar1,
  bar2,
  bar3,
 )
import Foo (foo)

singleElementList :: [String]
singleElementList = ["one string"]

multipleElementsList :: [String]
multipleElementsList =
  [ "one"
  , "two"
  , "three"
  ]

```

**Comment by georgefst** [comment]
We could add an option for splitting all multi-element lists, but personally I don't see it being flexible enough to be of much use.

Anyway, I've been thinking about this and it might actually be a much simpler change than I thought. To a first approximation, what we want to do is just hack [spansLayout](https://github.com/parsonsmatt/fourmolu/blob/cce6ecd2b556539019cf07d6d080195f4df7f76f/src/Ormolu/Printer/Combinators.hs#L132) to return `MultiLine` when one of the lines in that `SrcSpan` is too long. If anyone wants to experiment with that and open a PR, that'd be awesome. Otherwise I'll probably get to it in a few weeks.

**Comment by expipiplus1** [comment]
> Anyway, I've been thinking about this and it might actually be a much simpler change than I thought. To a first approximation, what we want to do is just hack [spansLayout](https://github.com/parsonsmatt/fourmolu/blob/cce6ecd2b556539019cf07d6d080195f4df7f76f/src/Ormolu/Printer/Combinators.hs#L132) to return `MultiLine` when we have a single line span that's too long. If anyone wants to experiment with that and open a PR, that'd be awesome. Otherwise I'll probably get to it in a few weeks.

This might not be trivial as the following input has a span length greater than 80 chars but will easily fit onto a single line

```haskell
foo = foo                                                                               bar
```

Perhaps one could get away with ignoring repeated space? Not sure how this would interact with the alignment stuff I'm working on though. 

I think in general the heuristic is right, in that one wants to make things multiline as far up the AST as possible.

**Comment by gustavoavena** [comment]
> We could add an option for splitting all multi-element lists, but personally I don't see it being flexible enough to be of much use.
> 
> Anyway, I've been thinking about this and it might actually be a much simpler change than I thought. To a first approximation, what we want to do is just hack [spansLayout](https://github.com/parsonsmatt/fourmolu/blob/cce6ecd2b556539019cf07d6d080195f4df7f76f/src/Ormolu/Printer/Combinators.hs#L132) to return `MultiLine` when we have a single line span that's too long. If anyone wants to experiment with that and open a PR, that'd be awesome. Otherwise I'll probably get to it in a few weeks.

I gave it a try with some small changes and this initial approach didn't work very well...

@georgefst is this more or less what you meant?
```
-- | Which layout combined spans result in?
spansLayout :: [SrcSpan] -> Layout
spansLayout = \case
  [] -> SingleLine
  (x : xs) ->
    if isOneLineSpan combinedSpan && not (shouldBreakSingleLine combinedSpan)
      then SingleLine
      else MultiLine
    where
      combinedSpan =  (foldr combineSrcSpans x xs)
      shouldBreakSingleLine :: SrcSpan -> Bool
      shouldBreakSingleLine (RealSrcSpan rs) =
          trace ("spanLineLength = " <> show spanLineLength) spanLineLength > 80
          where
              spanLineLength = srcSpanEndCol rs - srcSpanStartCol rs
      shouldBreakSingleLine _ = False
```

Original code:
```
_oneVeryLongList :: [String]
_oneVeryLongList = ["akjsndjklansdsad"] ++ ["jkanskdjnajsndjnasd"] ++ ["jknasdljknasdlnasdn"] ++ ["lajndljnasdlnasds"]
```

Formatted code:
```
_oneVeryLongList :: [String]
_oneVeryLongList =
  ["akjsndjklansdsad"] ++ ["jkanskdjnajsndjnasd"] ++ ["jknasdljknasdlnasdn"]
    ++ ["lajndljnasdlnasds"]
```

**Comment by expipiplus1** [comment]
> _oneVeryLongList

Looks more like several short lists to me...

**Comment by gustavoavena** [comment]
> Looks more like several short lists to me...

LOL, I think I was experimenting with this file a while ago and forgot to rename it when I changed to these lists... 😂

Either way, I think that it might actually be working well... I added a test case with an **actual long list** and it worked as intended: 

**Original code**:
```haskell

oneVeryLongLine :: [String]
oneVeryLongLine = ["akjsndjklansdsad"] ++ ["jkanskdjnajsndjnasd"] ++ ["jknasdljknasdlnasdn"] ++ ["lajndljnasdlnasds"]

oneVeryLongList :: [String]
oneVeryLongList = ["akjsndjklansdsad", "jkanskdjnajsndjnasd", "jknasdljknasdlnasdn", "lajndljnasdlnasds"]
```

**Formatted with my changes**
```haskell
oneVeryLongLine :: [String]
oneVeryLongLine =
  ["akjsndjklansdsad"] ++ ["jkanskdjnajsndjnasd"] ++ ["jknasdljknasdlnasdn"]
    ++ ["lajndljnasdlnasds"]

oneVeryLongList :: [String]
oneVeryLongList =
  [ "akjsndjklansdsad"
  , "jkanskdjnajsndjnasd"
  , "jknasdljknasdlnasdn"
  , "lajndljnasdlnasds"
  ]

```

**Formatted with current fourmolu release (nothing changes)**
```haskell
oneVeryLongLine :: [String]
oneVeryLongLine = ["akjsndjklansdsad"] ++ ["jkanskdjnajsndjnasd"] ++ ["jknasdljknasdlnasdn"] ++ ["lajndljnasdlnasds"]

oneVeryLongList :: [String]
oneVeryLongList = ["akjsndjklansdsad", "jkanskdjnajsndjnasd", "jknasdljknasdlnasdn", "lajndljnasdlnasds"]
```

Of course this is no guarantee that this works perfectly. I'll have to add many tests to make sure it doesn't break anything, but I think that maybe this is good enough to start working on a PR.

To make it simple, I'm planning on adding an option in `PrinterOptions` to set a line length limit, something like 
`poMaxLineLength :: f Maybe Int`, which would enforce this limit if the users sets a value. Otherwise just keep the current behaviour.

Any thoughts? 🙂

**Comment by expipiplus1** [comment]
One of the things which I'd most value would be for fourmolu to make "obviously single line" decisions. 

like changing this

```haskell
foo = 
  bar baz
```

into 

```haskell
foo = bar baz
```

Does your existing work do this?

**Comment by gustavoavena** [comment]
> Does your existing work do this?

Just tried it and it didn't change anything, so nope... But I wouldn't expect it to in this case. The only change I made was the modified `spansLayout` I posted in a comment above, which would only make some `SingleLine` layouts become `MultiLine`, but not the other way around.

**Comment by expipiplus1** [comment]
I quite often end up with single-line-fitting expressions after some text editor shenanigans (swapping arguments by swapping lines or some such). If it's not a huge trouble to do so perhaps that could be implemented?

This would probably need to take into account `scColumn` which in #94 becomes a little more difficult to use (as the expression layout can not be strict in it). Might require some thinking to get these two features to behave nicely with one another.

**Comment by pepeiborra** [comment]
Count me interested in this feature too, this is the reason I am still using Brittany.

> I quite often end up with single-line-fitting expressions after some text editor shenanigans (swapping arguments by swapping lines or some such). If it's not a huge trouble to do so perhaps that could be implemented?
> 
> This would probably need to take into account `scColumn` which in #94 becomes a little more difficult to use (as the expression layout can not be strict in it). Might require some thinking to get these two features to behave nicely with one another.

Joe, I think what @gustavoavena and @georgefst are proposing is a very simple hack: turn on the multi-line mode whenever the line is longer than N columns. It sounds like you are proposing something more fundamental that involves changes to the multi-line logic, and as such it belongs in a new issue and probably in the ormolu issue tracker

**Comment by expipiplus1** [comment]
Sure, I get that. I think as a first approximation for line-shortening would be the hack: turn on single-line mode whenever the line is shorter than `N - scColumn` columns.

**Comment by expipiplus1** [comment]
Although really what needs to happen would be some speculative layouting. Capturing and resetting the state after some speculative output should be fairly easy, but care might need to be taken to avoid the formatter becoming exponential in expression depth...

**Comment by brandonchinn178** [comment]
Implemented in #305

**Comment by gustavoavena** [comment]
Thanks for picking this up and finishing it @CrystalSplitter!
